packages/adblocker/src/resources.ts

```diff
@@ -36,129 +65,212 @@ export interface Resource {
+
+    return `data:${redirect.contentType},${redirect.content}`;
+  }
 
-    let dataUrl;
-    if (contentType.indexOf(';') !== -1) {
-      dataUrl = `data:${contentType},${body}`;
-    } else {
-      dataUrl = `data:${contentType};base64,${btoaPolyfill(body)}`;
+  public getRedirect(name: string): RedirectDetail {
+    const detail = this.redirectDetails.get(name);
+    if (detail !== undefined) {
+      return detail;
     }
 
-    return { body, contentType, dataUrl };
+    // Fallback to sample resources
+    const sample = getResourceForMime(name);
+    const dataUrl = sample.contentType.includes(';')
+      ? `data:${sample.contentType};base64,${btoaPolyfill(sample.body)}`
+      : `data:${sample.contentType},${sample.body}`;
+
+    return {
+      body: sample.body,
+      contentType: sample.contentType,
+      dataUrl,
+    };
   }
 
   public getSerializedSize(): number {
-    let estimatedSize = sizeOfASCII(this.checksum) + 2 * sizeOfByte(); // resources.size
+    let estimatedSize = sizeOfASCII(this.checksum);
 
-    this.resources.forEach(({ contentType, body }, name) => {
-      estimatedSize += sizeOfASCII(name) + sizeOfASCII(contentType) + sizeOfUTF8(body);
-    });
+    for (const redirect of new Set(this.redirects.values())) {
+      estimatedSize += redirect.names.reduce((state, name) => state + sizeOfASCII(name), 4);
+
+      estimatedSize += sizeOfUTF8(redirect.content);
+      estimatedSize += sizeOfASCII(redirect.contentType);
+      estimatedSize += 1;
+    }
+
+    for (const scriptlet of new Set(this.scriptlets.values())) {
+      estimatedSize += scriptlet.names.reduce((state, name) => state + sizeOfASCII(name), 4);
+
+      estimatedSize += sizeOfUTF8(scriptlet.content);
+      estimatedSize += scriptlet.dependencies.reduce(
+        (state, name) => state + sizeOfASCII(name),
+        4,
+      );
+      estimatedSize += 2;
+    }
 
     return estimatedSize;
   }
 
   public serialize(buffer: StaticDataView): void {
-    // Serialize `checksum`
     buffer.pushASCII(this.checksum);
 
-    // Serialize `resources`
-    buffer.pushUint16(this.resources.size);
-    this.resources.forEach(({ contentType, body }, name) => {
-      buffer.pushASCII(name);
-      buffer.pushASCII(contentType);
-      buffer.pushUTF8(body);
-    });
+    const redirects = new Set(this.redirects.values());
+    buffer.pushUint32(redirects.size);
+    for (const redirect of redirects) {
+      buffer.pushUint32(redirect.names.length);
+      for (const name of redirect.names) {
+        buffer.pushASCII(name);
+      }
+
+      buffer.pushUTF8(redirect.content);
+      buffer.pushUTF8(redirect.contentType);
+
+      switch (redirect.encoding) {
+        case 'base64': {
+          buffer.pushUint8(RedirectResourceEncodingMapping.BASE64);
+          break;
+        }
+        default: {
+          buffer.pushUint8(RedirectResourceEncodingMapping.RAW);
+        }
+      }
+    }
+
+    const scriptlets = new Set(this.scriptlets.values());
```

Comment:
instead of optimised data structure for scriptlets, cannot we simply `JSON.stringify` them?